### PR TITLE
Macos alt activator

### PR DESCRIPTION
**Pull request by gspencergoog**
## Description

This changes the semantics of supplying `alt: true` to `CharacterActivator` on macOS and iOS only so that specifying `alt: true` means to compare the supplied character with the `characterWithoutModifiers` field on the raw key data as opposed to the `character`.  This is because on these platforms, the Option (a.k.a. Alt) key modifies the character produced from the keyboard.  The expands the usefulness of the `CharacterActivator` so that you can, for instance, detect the `ß` character directly (so that pressing "option-s" on a US QWERTY keyboard, or pressing 'ß' on a German keyboard triggers it), or you can insist that the key be a combination of the Option key and the 's' key, so it only matches the US keystroke, but not the German one.

The need to support this comes from needing to support accelerator keys in menus drawn by Flutter. Since accelerators are derived from the characters in the localized label, but operate only when the option key is down, there needs to be a way to look for the unmodified character from the label while the option key is down. Otherwise it's impossible to, for instance, define an accelerator that triggers on an "s" keypress (where the "s" comes from the letter after the "&" in a label), since having the Option key down will cause it to produce a "ß" instead.

In order to simulate these key presses in tests, I had to add an additional parameter to the key simulation functions to pass the unmodified character. I cried a little when I did this, because I know that there is no equivalent for unmodified characters in `ui.KeyData` key events. We eventually will need to add some way to handle transmitting unmodified characters there, but I didn't address that with this PR.

## Related Issues
 - https://github.com/flutter/flutter/issues/23600

## Tests
 - Added tests for option key handling on macOS and iOS.

### CODE REVIEW

**Comment by dkwingsmt** [comment]
I'm deeply concerned how this change (and the `CharacterActivator::alt` change as a whole) should pan out as we transition into `KeyEvent`, which will not provide detailed `characters` and `charactersWithoutModifiers` but only `logicalKey` and `character`.

IIRC, on macOS, Option-S produces `logicalKey: .keyS` and `character: 'ß'`. Should we revert `CharacterActivator::alt` and make all `Alt-LetterKey` shortcuts use `SingleActivator` instead?

**Comment by gspencergoog** [comment]
Yep, I'm concerned too, but I don't see an alternative yet.

> Should we revert CharacterActivator::alt and make all Alt-LetterKey shortcuts use SingleActivator instead?

I can't: there's no mapping from a character to a `LogicalKeyboardKey`, and the accelerator triggers need to come from localized strings, where the translators have no knowledge of Dart or Flutter.

**Comment by gspencergoog** [comment]
After talking with Hans, we decided that we just won't support menu accelerators on macOS. The native menus don't support it, so it would be odd anyhow. Instead users will press "Ctrl-F2" to focus the menu, and then navigate with arrow keys after that. So, I could just close this, but...

It is a little weird on macOS to use the `alt:`: to use it is kind of redundant with the character produced. (e.g. `CharacterActivator('ß', alt: true)` is the same as `CharacterActivator('ß')`, and `CharacteActivator('s', alt: true)` will never fire). So we may still need something like this PR.

We still need support for `alt:` on other platforms, regardless.

**Comment by dkwingsmt** [comment]
I just realized I haven't followed up on this PR yet. Sorry for the delay.

LGTM.

File: packages/flutter/lib/src/widgets/shortcuts.dart [hunk 1]
```diff
@@ -607,8 +610,21 @@ class CharacterActivator with Diagnosticable, MenuSerializableShortcut implement
...
+  ///
+  /// On macOS and iOS, setting [alt] to true means that the comparison with the
+  /// character will use the unmodified character, since on these platforms
+  /// pressing the Option (⌥) key modifies the character that is produced. If
+  /// the intention is to match the character itself, and not the key
+  /// combination of ⌥+[character], then set [alt] to false and supply the
+  /// desired [character].
+  ///
+  /// For example, on a macOS US QWERTY keyboard you can either use
+  /// `CharacterActivator('ß')` or `CharacterActivator('s', alt: true)` to
```
Comment (dkwingsmt) [inline packages/flutter/lib/src/widgets/shortcuts.dart:623]: Are these docs out of date?

File: packages/flutter/lib/src/widgets/shortcuts.dart [hunk 2]
```diff
@@ -666,13 +682,27 @@ class CharacterActivator with Diagnosticable, MenuSerializableShortcut implement
   bool accepts(RawKeyEvent event, RawKeyboard state) {
     final Set<LogicalKeyboardKey> pressed = state.keysPressed;
     return event is RawKeyDownEvent
-      && event.character == character
+      && _characterMatches(event)
```
Comment (dkwingsmt) [inline packages/flutter/lib/src/widgets/shortcuts.dart:685]: Revert these changes?

Comment (gspencergoog) [reply packages/flutter/lib/src/widgets/shortcuts.dart:685]: If I revert these changes, I will just close the PR, since this is the only substantive change in the PR.

The question is: are we OK with the `alt:` behavior being counterintuitive on macOS and iOS?
Using `alt:` right now is kind of redundant with the character produced on those platforms (e.g. `CharacterActivator('ß', alt: true)` is the same as `CharacterActivator('ß')`, and `CharacteActivator('s', alt: true)` will never fire).

The alternative is this code, which will use the character without modifier to allow `CharacterActivator('s', alt: true)` to fire when the non-modifier character is 's', and Option is down, making things a little less surprising. I realize that this means that the new event system will need to support an optional `characterWithoutModifier` field, but here is a use case where that is appropriate, and I don't see any other way to do the job on macOS and iOS.

If a program wants to generate its shortcuts from a string source (e.g. in the case of accelerators), and wants to know if the Alt/Option key is also down at the same time, then they have no way to implement that currently on macOS and iOS, since the option-key mapping of the keys is different for different keyboard layouts.

Luckily, for my use case I can just refuse to implement accelerators on macOS, but if someone wants to implement something that works like an accelerator they'll be unable to under the new key event system (as it is now, they could conceivably create their own activator with code like the code here, but once the old key event system is gone, they won't be able to).

**Comment by gspencergoog** [comment]
I'm going to close this until we can get some consensus, and a clearer use case shows itself.
